diff --git a/src/cumesh.h b/src/cumesh.h
index c1754d1..336d724 100644
--- a/src/cumesh.h
+++ b/src/cumesh.h
@@ -452,7 +452,13 @@ public:
      * @param threshold The threshold for edge collapse cost.
      * @return A tuple of the number of vertices and the number of faces after simplification.
      */
-    std::tuple<int, int> simplify_step(float lambda_edge_length, float lambda_skinny, float threshold, bool timing=false);
+    std::tuple<int, int> simplify_step(
+        float lambda_edge_length,
+        float lambda_skinny,
+        float threshold,
+        bool timing=false,
+        bool reuse_vertex_face_adjacency=false
+    );
 
 
     // Atlasing functions
diff --git a/src/ext.cpp b/src/ext.cpp
index 92b86de..f2c7f35 100644
--- a/src/ext.cpp
+++ b/src/ext.cpp
@@ -58,7 +58,10 @@ PYBIND11_MODULE(TORCH_EXTENSION_NAME, m) {
         .def("remove_non_manifold_faces", &cumesh::CuMesh::remove_non_manifold_faces)
         .def("remove_small_connected_components", &cumesh::CuMesh::remove_small_connected_components)
         .def("unify_face_orientations", &cumesh::CuMesh::unify_face_orientations)
-        .def("simplify_step", &cumesh::CuMesh::simplify_step)
+        .def("simplify_step", &cumesh::CuMesh::simplify_step,
+             py::arg("lambda_edge_length"), py::arg("lambda_skinny"),
+             py::arg("threshold"), py::arg("timing") = false,
+             py::arg("reuse_vertex_face_adjacency") = false)
         .def("compute_charts", &cumesh::CuMesh::compute_charts)
         .def("read_atlas_charts", &cumesh::CuMesh::read_atlas_charts);
 
diff --git a/src/simplify.cu b/src/simplify.cu
index 1d30d4a..c8daae1 100644
--- a/src/simplify.cu
+++ b/src/simplify.cu
@@ -520,11 +520,19 @@ void collapse_edges(
 }
 
 
-std::tuple<int, int> CuMesh::simplify_step(float lambda_edge_length, float lambda_skinny, float threshold, bool timing) {
+std::tuple<int, int> CuMesh::simplify_step(
+    float lambda_edge_length,
+    float lambda_skinny,
+    float threshold,
+    bool timing,
+    bool reuse_vertex_face_adjacency
+) {
     std::chrono::high_resolution_clock::time_point start, end;
 
     if (timing) start = std::chrono::high_resolution_clock::now();
-    this->get_vertex_face_adjacency();
+    if (!reuse_vertex_face_adjacency) {
+        this->get_vertex_face_adjacency();
+    }
     if (timing) {
         CUDA_CHECK(cudaDeviceSynchronize());
         end = std::chrono::high_resolution_clock::now();
